metadata_please/source_checkout.py

```diff
@@ -0,0 +1,237 @@
+    """
+    try:
+        data = (path / "pyproject.toml").read_text()
+    except FileNotFoundError:
+        return b""
+    doc = toml.loads(data)
+
+    buf: list[str] = []
+    for dep in doc.get("project", {}).get("dependencies", ()):
+        buf.append(f"Requires-Dist: {dep}\n")
+    for k, v in doc.get("project", {}).get("optional-dependencies", {}).items():
+        extra_name = canonicalize_name(k)
+        buf.append(f"Provides-Extra: {extra_name}\n")
+        for i in v:
+            buf.append("Requires-Dist: " + merge_extra_marker(extra_name, i) + "\n")
+
+    return "".join(buf).encode("utf-8")
+
+
+def _translate_caret(specifier: str) -> str:
+    """
+    Given a string like "^0.2.3" returns ">=0.2.3,<0.3.0".
+    """
+    assert "," not in specifier
+    parts = specifier[1:].split(".")
+    while len(parts) < 3:
+        parts.append("0")
+
+    for i in range(len(parts)):
+        if parts[i] != "0":
+            # The docs are not super clear about how this behaves, but let's
+            # assume integer-valued parts and just let the exception raise
+            # otherwise.
+            incremented = parts[:]
+            incremented[i] = str(int(parts[i]) + 1)
+            del incremented[i + 1 :]
+            incremented_version = ".".join(incremented)
+            break
+    else:
+        raise ValueError("All components were zero?")
+    return f">={specifier[1:]},<{incremented_version}"
+
+
+def _translate_tilde(specifier: str) -> str:
+    """
+    Given a string like "~1.2.3" returns ">=1.2.3,<1.3".
+    """
+    assert "," not in specifier
+    parts = specifier[1:].split(".")
+    incremented = parts[:2]
+    incremented[-1] = str(int(incremented[-1]) + 1)
+    incremented_version = ".".join(incremented)
+
+    return f">={specifier[1:]},<{incremented_version}"
+
+
+def from_poetry_checkout(path: Path) -> bytes:
+    """
+    Returns a metadata snippet (which is zero-length if this is none of this style).
+    """
+    try:
+        data = (path / "pyproject.toml").read_text()
+    except FileNotFoundError:
+        return b""
+    doc = toml.loads(data)
+
+    saved_extra_constraints = {}
+
+    buf: list[str] = []
+    for k, v in doc.get("tool", {}).get("poetry", {}).get("dependencies", {}).items():
+        if k == "python":
+            pass  # TODO requires-python
+        else:
+            k = canonicalize_name(k)
+            if isinstance(v, dict):
+                version = v.get("version", "")
+                if "extras" in v:
+                    extras = "[%s]" % (",".join(v["extras"]))
+                else:
+                    extras = ""
+                markers = v.get("markers", "")
+                python = v.get("python", "")
+                if python:
+                    m = OPERATOR_RE.fullmatch(python)
+                    assert m is not None
+                    # TODO do ^/~ work on python version?
```

Comment:
I assume this goes through the same logic as generic version matching, so I should hoist the ^/~ handling out into something I can call here.  Failing tests would be good.